codec/src/main/java/io/netty/handler/codec/compression/JdkZlibEncoder.java

```diff
@@ -198,6 +260,56 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf uncompressed, ByteBuf o
             return;
         }
 
+        deflaterSetInput(uncompressed, out, len);
+        for (;;) {
+            deflate(out);
+            if (deflater.needsInput()) {
+                // Consumed everything
+                break;
+            } else {
+                if (!out.isWritable()) {
+                    // We did not consume everything but the buffer is not writable anymore. Increase the capacity to
+                    // make more room.
+                    out.ensureWritable(out.writerIndex());
+                }
+            }
+        }
+    }
+
+    private void deflaterSetInput(ByteBuf uncompressed, ByteBuf out, int len) {
+        if (DEFLATER_SET_INPUT_BYTEBUFFER != null && CRC32_UPDATE_BYTEBUFFER != null) {
+            deflaterSetInputJdk11(uncompressed, out, len);
+        } else {
+            deflaterSetInputJdk6(uncompressed, out, len);
+        }
+    }
+
+    private void deflaterSetInputJdk11(ByteBuf uncompressed, ByteBuf out, int len) {
+        if (uncompressed.isDirect()) {
+            if (writeHeader) {
+                writeHeader = false;
+                if (wrapper == ZlibWrapper.GZIP) {
+                    out.writeBytes(gzipHeader);
+                }
+            }
+
+            try {
+                if (wrapper == ZlibWrapper.GZIP) {
+                    CRC32_UPDATE_BYTEBUFFER.invoke(crc, uncompressed.nioBuffer());
+                }
+
+                DEFLATER_SET_INPUT_BYTEBUFFER.invoke(deflater, uncompressed.nioBuffer());
```

Comment:
nit: would it make sense to store the `ByteBuffer` in a variable and so reduce GC ?

Something like:

```java
ByteBuffer nioBuffer = uncompressed.nioBuffer();
if (wrapper == ZlibWrapper.GZIP) {
    int position = nioBuffer.position();
    int limit = nioBuffer.limit();
    CRC32_UPDATE_BYTEBUFFER.invoke(crc, nioBuffer);
    nioBuffer.position(position).limit(limit);
}
DEFLATER_SET_INPUT_BYTEBUFFER.invoke(deflater, nioBuffer());

```

Reply:
done